cpp/log_rotate.cc

```diff
@@ -98,10 +101,17 @@ int Main(int argc, char *argv[]) {
         if (not HasNumericExtension(entry.getName())) {
             if (verbose)
                 std::cout << "About to rotate \"" << entry.getName() << "\".\n";
-            if (max_line_count > 0)
-                FileUtil::OnlyKeepLastNLines(directory_path + "/" + entry.getName(), max_line_count);
-            else {
-                const std::string filename(directory_path + "/" + entry.getName());
+
+            const std::string filename(directory_path + "/" + entry.getName());
+            if (max_line_count > 0) {
+                const int fd(::open(filename.c_str(), O_RDWR));
+                { // New scope to ensure that the FileLocker instance goes out of scope before we close fd!
+                    FileLocker file_locker(fd, FileLocker::WRITE_ONLY);
```

Comment:
FileUtil::OnlyKeepLastNLines opens it with O_RDONLY to calculate no_of_reading_bytes_to_remove, then it passes it on to ::truncate later which needs write access.

FileLocker only supports READ_ONLY and WRITE_ONLY, maybe we need a READ_WRITE flag?

Reply:
Or maybe WRITE_ONLY would be correct and should be renamed to READ_WRITE?

Reply:
FileUtil::OnlyKeepLastNLines calls RemoveLeadingBytes which opens the file with O_RDWR before calling ftruncate.
As far as the locking is concerned that is my reading of what the underlying fcntl documentation states.

Reply:
Again: Why is the flag called WRITE_ONLY if, in fact, we use it to both reading + writing?